cloudstack/loadbalancer.go

```diff
@@ -141,48 +155,65 @@ func (cs *CSCloud) GetLoadBalancer(ctx context.Context, clusterName string, serv
 	klog.V(4).Infof("Found a load balancer %v", lb)
 
 	status := &v1.LoadBalancerStatus{}
-	status.Ingress = append(status.Ingress, v1.LoadBalancerIngress{IP: lb.ip.address})
+	status.Ingress = append(status.Ingress, v1.LoadBalancerIngress{
+		IP:       lb.ip.address,
+		Hostname: lb.name,
+	})
 
 	return status, true, nil
 }
 
 // EnsureLoadBalancer creates a new load balancer, or updates the existing one. Returns the status of the balancer.
-func (cs *CSCloud) EnsureLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) (status *v1.LoadBalancerStatus, err error) {
-	klog.V(5).Infof("EnsureLoadBalancer(%v, %v, %v, %v, %v, %#v)", clusterName, service.Namespace, service.Name, service.Spec.LoadBalancerIP, service.Spec.Ports, nodes)
+func (cs *CSCloud) EnsureLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) (*v1.LoadBalancerStatus, error) {
+	if service == nil {
+		return nil, fmt.Errorf("EnsureLoadBalancer: service cannot be nil")
+	}
+
+	klog.V(5).Infof("EnsureLoadBalancer(%v, %v, %v, %v, ports: %d, nodes: %d)", clusterName, service.Namespace, service.Name, service.Spec.LoadBalancerIP, len(service.Spec.Ports), len(nodes))
 	cs.svcLock.Lock(service)
 	defer cs.svcLock.Unlock(service)
 
 	if len(service.Spec.Ports) == 0 {
 		return nil, fmt.Errorf("requested load balancer with no ports")
 	}
 
-	nodes = cs.filterNodesMatchingLabels(nodes, *service)
-
-	if len(nodes) == 0 {
-		return nil, fmt.Errorf("no nodes available to add to load balancer")
-	}
+	nodes = cs.filterNodesMatchingLabels(nodes, service)
 
 	hostIDs, networkIDs, projectID, err := cs.extractIDs(nodes)
 	if err != nil {
 		return nil, err
 	}
-	if len(hostIDs) == 0 || len(networkIDs) == 0 {
-		return nil, fmt.Errorf("unable to map kubernetes nodes to cloudstack instances and networks for service (%v, %v)", service.Namespace, service.Name)
-	}
 
 	// Get the load balancer details and existing rules.
 	lb, err := cs.getLoadBalancer(service, projectID, networkIDs)
 	if err != nil {
 		return nil, err
 	}
 
-	lb.setAlgorithm(service)
+	if lb.cloud.projectID != "" && service.Labels[cs.config.Global.ProjectIDLabel] == "" {
+		service.Labels[cs.config.Global.ProjectIDLabel] = lb.cloud.projectID
+
+		_, err = cs.kubeClient.CoreV1().Services(service.Namespace).Patch(
```

Comment:
I checked the service controller code and it also uses PATCH to update the service status, so incrementing the ResourceVersion here is not a problem.